site/source/docs/porting/simd.rst

```diff
@@ -35,6 +35,53 @@ When porting native SIMD code, it should be noted that because of portability co
 
 SIMD-related bug reports are tracked in the `Emscripten bug tracker with the label SIMD <https://github.com/emscripten-core/emscripten/issues?q=is%3Aopen+is%3Aissue+label%3ASIMD>`_.
 
+===========================
+Optimization considerations
+===========================
+
+When porting SIMD code to use WebAssembly SIMD, implementors should be aware of semantic differences between the host hardware and WebAssembly semantics; as acknowledged in the WebAssembly design documentation, "`this sometimes will lead to poor performance <https://github.com/WebAssembly/design/blob/master/Portability.md#assumptions-for-efficient-execution>`_." The following list outlines some WebAssembly SIMD instructions to look out for when performance tuning:
+
+.. list-table:: WebAssembly SIMD instructions with performance implications
+   :widths: 10 10 30
+   :header-rows: 1
+
+   * - WebAssembly SIMD instruction
+     - Hardware architecture
+     - Considerations
+
+   * - [i8x16|i16x8|i32x4|i64x2].[shl|shr_s|shr_u]
+     - x86, arm
+     - Use a constant shift amount to avoid extra instructions checking that it is in bounds.
+
+   * - i8x16.[shl|shr_s|shr_u]
+     - x86
+     - Included for orthogonality, these instructions have no equivalent x86 instruction and are emulated with 5-11 x86 instructions (i.e. using 16x8 shifts).
```

Comment:
This is incorrect when the amount is constant (... and if it's correct we need to file v8 bugs :D) - it's still emulated but should just be 2-3 insns.

Reply:
I think it's still 5 when the amount is constant: https://github.com/v8/v8/blob/cfdac7f91d1fc444b5e0989f7f56ef72a3b2c80f/src/compiler/backend/x64/code-generator-x64.cc#L3444-L3506. What is the 2-3 instruction lowering you would expect? I guess I'm counting moves in there...

Reply:
I was hoping for psslw + pand with a memory location or psslw+movaps+and. Looks like it takes 3 instructions just to build the constant up; I don't think any of these moves are free either. Not sure if there's a v8 bug for this yet but definitely this could be better.

Reply:
("free" as in reg-renamed - I believe only moves between the same register file and zeroes are free)

Reply:
Doing what you suggest requires a way to write out some masks in memory and then reference them. I added a constant pool mechanism to Cranelift for this reason (and you can see the shift masks [here](https://github.com/bytecodealliance/wasmtime/blob/dfffc6915026f1df904162bc82eb8bfd49e810f9/cranelift/codegen/src/isa/x64/lower.rs#L2162-L2195)) but V8 does not have this yet. @omnisip is working on a [constant pool implementation in V8](https://chromium-review.googlesource.com/c/v8/v8/+/2798529) that would help solve this. Perhaps we should add this shift use case to the [design document](https://docs.google.com/document/d/1W40WUTcLPYE3uBek_olTgiOgyKXMr4BP3o8m9uzv6Y8/edit?usp=sharing)?